File: src/main/java/co/onmind/microhex/infrastructure/notification/NotificationAdapter.java
package co.onmind.microhex.infrastructure.notification;

import co.onmind.microhex.domain.models.Role;
import co.onmind.microhex.domain.ports.out.NotificationPort;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

/**
 * Implementation of the NotificationPort for sending notifications
 * about role operations to external systems.
 * 
 * This is a simple implementation that logs notifications.
 * In a real-world scenario, this could send messages to message queues,
 * webhooks, email services, etc.
 */
@Component
public class NotificationAdapter implements NotificationPort {
    
    private static final Logger logger = LoggerFactory.getLogger(NotificationAdapter.class);
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void notifyRoleCreated(Role role) {
        logger.info("NOTIFICATION: Role created - ID: {}, Name: {}, CreatedAt: {}", 
                   role.getId(), role.getName(), role.getCreatedAt());
        
        // Here you could implement actual notification logic:
        // - Send message to message queue (RabbitMQ, Apache Kafka, etc.)
        // - Send webhook notification
        // - Send email notification
        // - Update external systems
        // - Trigger other business processes
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void notifyRoleUpdated(Role role) {
        logger.info("NOTIFICATION: Role updated - ID: {}, Name: {}, CreatedAt: {}", 
                   role.getId(), role.getName(), role.getCreatedAt());
        
        // Here you could implement actual notification logic for updates
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void notifyRoleDeleted(Long roleId) {
        logger.info("NOTIFICATION: Role deleted - ID: {}", roleId);
        
        // Here you could implement actual notification logic for deletions
    }
}

File: src/main/kotlin/co/onmind/hex/infrastructure/notification/adapter/NotificationAdapter.kt
package co.onmind.hex.infrastructure.notification.adapter

import co.onmind.hex.domain.model.Role
import co.onmind.hex.domain.ports.out.NotificationPort
import jakarta.inject.Singleton
import org.slf4j.LoggerFactory

/**
 * Adapter that implements the NotificationPort using logging.
 * 
 * This is a simple implementation that logs notifications.
 * In a real application, this could send emails, push notifications,
 * or integrate with external messaging systems.
 */
@Singleton
class NotificationAdapter : NotificationPort {
    
    private val logger = LoggerFactory.getLogger(NotificationAdapter::class.java)
    
    override fun notifyRoleCreated(role: Role) {
        logger.info("Role created notification: Role '${role.name}' with ID ${role.id} was created at ${role.createdAt}")
    }
    
    override fun notifyRoleUpdated(role: Role) {
        logger.info("Role updated notification: Role '${role.name}' with ID ${role.id} was updated")
    }
    
    override fun notifyRoleDeleted(roleId: Long) {
        logger.info("Role deleted notification: Role with ID $roleId was deleted")
    }
}